### DISCUSSION TITLE
New storage UI handling and generating the same config than unattended installation

### DISCUSSION THREAD

**Discussion by ancorgs** [agama-project/agama#1678] [Ideas]
## The overall idea

Let's start with a quick reminder of the current UI. It looks like this...

![current](https://github.com/user-attachments/assets/23a0034e-3ef8-4304-a91e-83c52bfc3363)

... and generates an object `ProposalSettings` with a relatively high level of abstraction. That is used as input for an algorithm relatively similar to the YaST Guided Setup.

On the other hand, when working on unattended mode, Agama takes an input an object `Storage::Config` that is described at auto_storage.md (see the [latest pull request](https://github.com/agama-project/agama/pull/1671)). That format allows to express everything that can be expressed via `ProposalSettings` and also many more things like creation of MD RAIDs.

We are considering to change the general approach of the UI to read and generate `Storage::Config` and completely kill the concept of `ProposalSettings`. In fact, that was the original concept a couple of years ago, but we deviated from it for several reasons.

Let's use this discussion to propose ideas and alternatives following that core principle: working with `Storage::Config` objects and always running the same algorithm of the unattended installation.

## Changing the approach to sizes

Each entry at the filesystems box in the current UI corresponds to the underlying concept of one volume of `ProposalSettings`. And it's possible to specify a size of "auto" for a filesystem (volume). That value can only be used on some volumes that support that. Because that "auto" flag only makes sense on the (pretty artificial) world of the volumes and means a volume can follow a relatively complex logic to calculate its size limits (adapted to RAM, grow if there is no separate home, etc.). For example, at Tumbleweed "/home" has a default max and min, but no "auto" because those size limits (min and max) are always the same.

The new interfaces will still have a list of filesystems in one way or another. But the entries will not be translated into volumes for `ProposalSettings`. Very likely each entry will represent an element `filesystem` of the `Storage::Config`, together with the corresponding partition/lv that will be created to host that filesystem.

That's also a nice opportunity to modify/kill the concept of "auto" in the UI.

At `Storage::Config` sizes for devices can be omitted and Agama will do the appropriate thing. From that point of view and following the previous example, omitting the sizes of "/home" is not different from omitting the size of a volume with automatic sizes. The real decision is "I will enter my own size (or size limits)" vs "I will let the installer decide on the size". How then the installer decides the sizes on that second case is important but secondary.

Take into account there are also situations in which specifying sizes simply makes no sense. Eg. when directly formatting/mounting a device instead of creating a new partition/lv.

So setting your own size limits for a given mount point feels more like an optional thing that can be done by the user in some situations. But it makes sense if the default at the UI is omitting the size and letting the installer decide.

**Comment by ancorgs** [comment]
## Idea 1. Moving one step from the current UI

“Installation device” box becomes “Installation devices” and allows to add:
  - Any disk
  - Any existing RAID
  - Any existing LVM (to be implemented later)
  - A new LVM defining existing block devices as PVs
  - A new LVM selecting a disk or RAID which will be automatically partitioned to accommodate the PVs
  - A new RAID defining existing block devices as members
  - A new set of RAIDs selecting some disks that will be partitioned to accommodate partitions that will then be grouped at RAIDs
  - Btrfs RAIDs may also be included. I have not considered them yet.

Every new device will receive an alias that can be always auto-calculated (based on the device name) but that the user can override.

Adding a new device to the list only adds that device, not necessarily the underlying ones (eg. a new LVM that will automatically create partitions on sda and sdb does not add those disks to the list).

We may need to reconsider some aspects of the filesystems box (currently titled "Partitions and filesystems").

Each entry can represent either:
  - a new partition over a disk/RAID (with its corresponding filesystem)
  - a new logical volume over an LVM (with its corresponding filesystem)
  - a filesystem directly located on a device
In the first two cases, that device will be represented at "installation devices" (that does not necessarily imply that the we need to add it as a mandatory previous step). I'm not so sure about the third case.

Using a device (instead of creating a new partition/lv) is probably something that should be more integrated into the edition of the filesystem, since it has important implications (eg. the sizes make no sense and maybe the "reuse" flags from `filesystem` and `encryption` become relevant).

The approach to sizes also change. See a separate comment about that.

**Comment by ancorgs** [comment]
## Idea 2. One more step from the current UI to the Expert Partitioner

This implies merging the "Installation devices" mentioned on idea 1 and the list of filesystems. The same kind of devices can be added to that list - ie. disks, volume groups (new or reused) and RAIDs (new or reused) - and each device still receives an alias.

But many other aspects are different from idea 1

Adding a device implies adding also the underlying ones. Eg. if a new LVM is defined with automatically generated PVs as partitions on "sdb", then both the new VG and sdb will be displayed on the list. Even if sdb is used only to host those automatic PVs and nothing else.

The disk selected for booting (if any) is also part of the list. Even if it's not used for any other purpose.

The devices in the list can be directly formatted or contain a list of partitions/lv. In both cases, that information is displayed next to its corresponding device. Let's imagine the system is configured to boot from sda but the filesystems will be created as new logical volumes on top of a new LVM with auto-generated partitions on sdb. And to make things a bit more interesting, let's say a previously existing partition from sdb will be reused. Conceptually that would look like (of course, the representation may vary a lot from what can be expressed with HTML unordered lists).

- sda. Partitioned disk.
  - Existing partitions will be kept.
  - System partitions:
    - _Partitions needed for booting_
- sdb. Partitioned disk.
  - One existing partition could be shrunk.
  - System Partitions:
    - _New physical volume(s) for vg_system_
    - Existing sdb2 mounted at /old_home
- vg_system. New LVM.
  - System Logical volumes:
    - New / btrfs with snapshots
    - New swap

The entries in italic represent aspects that are defined indirectly (eg. using the boot settings or creating a new generated volume group on top of the device). In any case, all the information in the list above can be directly obtained from `Storage::Config`. It does not depend on the execution of the proposal.

The actions to make space on existing devices (what to resize/delete) is defined by device instead of globally. The option is accessed from each entry at the above list. In the case of directly formatted devices the content is always deleted so there is nothing to influence (although the information is still displayed). Eg.

- sda. Disk.
  - Existing content will be deleted
  - _Member of md0_
- sdb. Disk.
  - Existing content will be deleted.
  - _Member of md0_
- md0. New RAID.
  - System Partitions:
    - _Partitions needed for booting_
    - New / btrfs with snapshots
    - New swap

Note the previous example could end up placing `/boot/efi` (or any other boot partition) inside a RAID because md0 has been selected as boot device at the corresponding boot settings. That's something YaST users do occasionally. It's allowed by the Expert Partitioner displaying [this warning](https://github.com/yast/yast-storage-ng/blob/fdc1614cbe438001b4729475ea2034f6f5ca0400/src/lib/y2storage/boot_requirements_strategies/uefi.rb#L80).

It would make sense to have a way to specify "default" as part of the list of partitions for a disk. Then the UI could generate a config like this:

```json
{ "storage": {
  "disks": [
    { "search": "/dev/sda", "partitions": ["default"] }
  ]
} }
```

That would look like this

- sda. Partitioned Disk.
  - Existing partitions will be kept.
  - System Partitions:
    - _Partitions needed for booting_
    - _Default partitions_

Even without having executed the proposal yet, that information can actually be already expanded to a so-called "resolved" object `Storage::Config` and then displayed as something like this.

- sda. Partitioned Disk.
  - Existing partitions will be kept.
  - System Partitions:
    - _Partitions needed for booting_
    - _New default / btrfs with snapshots_
    - _New default swap_

Note the root and swap partitions are italic because they are not explicitly specified, they are the consequence of selecting sda as the target for the default filesystems.

Probably the concept will not scale very well to represent something like this

```json
{ "storage": {
  "drives": [...],
  "mdRaids": [
    "generate": {
        "mdRaids": "default",
        "level": "raid0",
        "devices": [
          { "generate": ["sda", "sdb"] }
        ]
    }
  ]
} }
```

But still worth exploring.

Likely the settings for encryption could also be integrated in the same box. So all the settings (boot settings, drives, partitions, lvm...) will be defined in one box. The result could also be merged into a single box or kept as two separate ones for "Actions" and "Final layout". Take into account "Actions" would become less relevant because the so called SpaceActions are now configured for each device together with other aspects.

### To be discussed: displaying sizes at the config

The list above does not display the sizes of the proposed partitions or logical volumes.  That's debatable, but I think showing the size limits may be confusing.

Maybe an alternative would be to only display size limits when they are explicit, as opposed to the omitted ones that are actually the default case (or that's what we expect).

In any case, there must be ways to check which min and max sizes were used, no matter if explicitly set by the user or inferred by Agama. And in the second case (omitted sizes) the explanation for the used limits (eg. based on RAM) should also be presented. But maybe all that information should only be revealed when needed. For example, when editing a filesystem or when clicking on some "explain me this" button/hover.

**Reply by ancorgs** [reply]
## Idea 2bis.

This reply presents some possible modifications over idea 2.

Basically, it merges in the list of settings some information about how the final result will look like. We need to decide whether that helps to understand the settings or introduces more confusion.

From a technical point of view, it means the represented list of settings is not a direct representation of `Storage::Config`. It also includes information coming from the resulting list of actions (and maybe also the devicegraph).

The first example would turn into something like this:

- sda. Partitioned disk.
  - Existing partitions will be kept.
  - System partitions:
    - _Existing sda1 mounted at /boot/efi (512 MiB)_
- sdb. Partitioned disk.
  - One existing partition will be shrunk.
  - System Partitions:
    - _Two new physical volumes for vg_system (40 GiB)_
    - Existing sdb2 mounted at /old_home (15 GiB)
- vg_system. New LVM.
  - System Logical volumes:
    - New / btrfs with snapshots (30 GiB)
    - New swap (10 GiB)

The most obvious difference is the sizes. See also that the actions to make space are more specific, not only explaining what is configured but also exactly how many partitions are effectively affected. In addition, "partitions needed for booting" is actually replaced by a more concrete explanation of what will be done in that regard. Last but not least, the message about auto-generated PVs for vg_system could be more explicit indicating the number of PVs created on that concrete disk.

The second example would not change that much, but still would include several more details about final sizes, about deleting and about booting.

- sda. Disk.
  - Three existing partitions will be deleted
  - _Member of md0_
- sdb. Disk.
  - One existing partition will be deleted.
  - _Member of md0_
- md0. New RAID.
  - System Partitions:
    - _New /boot/efi (256MiB)_
    - New / btrfs with snapshots (300 GiB)
    - New swap (10 GiB)

The third example would be very similar to one of the alternatives presented at idea 2.

- sda. Partitioned Disk.
  - Existing partitions will be kept.
  - System Partitions:
    - _Existing sda1 mounted at /boot/efi (512 MiB)_
    - _New default / btrfs with snapshots (300 GiB)_
    - _New default swap (10 GiB)_

### To be discussed: displayed sizes

The sizes displayed in all the previous examples are the final ones after generating the proposal.

Everything mentioned at idea 2 about displaying (and explaining) the size limits (explicit or not) when editing the entry (and/or with some other mechanism) still applies.

**Reply by ancorgs** [reply]
Based on an offline conversation with @dgdavid, this would be a first (very quick) attempt to have collapsed items at the list of devices. I will put more time into it when I have more time.

**Full**

- sda. Partitioned Disk.
  - Existing partitions will be kept.
  - System Partitions:
    - _Partitions needed for booting_
    - _New default / btrfs with snapshots_
    - _New default swap_

**Collapsed**

- sda. Boot disk. New partitions will be created. Content will be kept.

"Boot disk" could even be an icon.

**Full**

- sda. Partitioned disk.
  - Existing partitions will be kept.
  - System partitions:
    - _Partitions needed for booting_
- sdb. Partitioned disk.
  - One existing partition could be shrunk.
  - System Partitions:
    - _New physical volume(s) for vg_system_
    - Existing sdb2 mounted at /old_home
- vg_system. New LVM.
  - System Logical volumes:
    - New / btrfs with snapshots
    - New swap

**Collapsed**

- sda. Boot disk. Content will be kept.
- sdb. Will host new LVM "system". Some partitions will be reused and other could be shrunk.
- New LVM "system". New logical volumes will be created.

**Full**

- sda. Disk.
  - Existing content will be deleted
  - _Member of md0_
- sdb. Disk.
  - Existing content will be deleted.
  - _Member of md0_
- md0. New RAID.
  - System Partitions:
    - _Partitions needed for booting_
    - New / btrfs with snapshots
    - New swap

**Collapsed**

- sda. Part of new RAID "md0". Content will be deleted.
- sdb. Part of new RAID "md0". Content will be deleted.
- New RAID "md0". Boot disk. Several partitions will be created.

**Reply by dgdavid** [reply]
Thanks a ton @ancorgs!

**Reply by joseivanlopez** [reply]
What about sharing this idea to a more wide audience? At least with the rest of the team. Maybe we gather some interesting feedback. Or at least to have a serious review after the first iteration with a "read only" interface.

**Reply by dgdavid** [reply]
Sure. We are working on it and writing it here and not keeping it private was a way of sharing.

@ancorgs not only wrote it here for publishing with some context, but also put the link in the IRC channel where usually the rest of the team is.

When we have a clear idea of the UI, we can publish it for gathering feedback. Meanwhile we are just trying to shape it. Hope you understand that involving too many people in that phase could be a bit challenging. Nevertheless, we are open to listen/read ideas if someone have one.

**Reply by joseivanlopez** [reply]
I meant about sharing the overall approach. There are some key elements in the "idea 2". For example, having two separate sections, one for defining the proposal config and other for seeing the result. That is not the "typical" approach for working with disks, partitions, etc. I would like to know what people think about it. I am fine guys if you think is too early for that. But even a very premature version of the UI could work for exposing the "key concepts" I mentioned before.

**Comment by joseivanlopez** [comment]
## Idea 3. Expert Partitioner with "Helpers"

Having something like the Expert Partitioner is perfectly doable thanks to the power of the *agama proposal* and the *storage config*.  

The UI would consist on a single table of devices, always showing the result. This is similar to what the current UI does in the *result* box.

![Screenshot from 2024-10-18 09-09-46](https://github.com/user-attachments/assets/04c6f254-28e1-4b17-9f9b-ccc7ff4b8dc3)

The UI offers a set of actions for each device:

* For a disk: remove content, create partition table, etc.
* For a partition: edit, resize, delete.
* For unpartitioned space: create partition.
* ...

Moreover, there would be some "advanced forms" (i.e., helpers) to easily perform certain actions. For example:

* A helper for creating all the partitions required by the product.
* A helper for easily creating a LVM volume group.
* A helper for creating all the logical volumes required by the product.
* A helper for creating a RAID by simply selecting the disks to use.
* Etc.

![Screenshot from 2024-10-18 09-10-06](https://github.com/user-attachments/assets/09ccedf4-6585-4f5f-8549-619e3ff737db)

From the implementation point of view, the UI does not need to understand the *storage config* in order to draw the table of devices. It simply uses the "devicegraph" (as the result does).

All the possible actions are already (or will be) supported by the *agama proposal*, so the UI only has to generate the proper *storage config* for the requested action. For example: 

* For creating a partition:

```json
{
  "drives": [
    {
      "search": "/dev/vda",
      "partitions": [
        {  
          "size": "10 GiB",
          "filesystem": {
            "type": "xfs",
            "path": "/home"
          }
        }
      ]
    }
  ]
}
```

* For deleting a partition:

```json
{
  "drives": [
    {
      "search": "/dev/vda",
      "partitions": [
        {  
          "search": "/dev/vda1",
          "delete": true
        }
      ]
    }
  ]
}
```

* For creating a volume group (with a helper):

```json
{
  "drives": [
    { 
      "search": "/dev/vda",
      "alias": "vda"
    },
    { 
      "search": "/dev/vdb",
      "alias": "vdb"
    }
  ]
  "volumeGroups": [
    {
      "physicalVolumes": [
        { "generate": ["vda", "vdb"] }
      ]
    }
  ]
}
```

* For creating a partition in a specific free slot: 

```json
{
  "drives": [
    {
      "search": "/dev/vda",
      "partitions": [
        {  
          "sectorStart": 12442323342,
          "size": "10 GiB",
          "filesystem": {
            "type": "xfs",
            "path": "/home"
          }
        }
      ]
    }
  ]
}
```

The *storage config* only needs to add a property in the partition schema (e.g., *sectorStart*) to indicate the number of the first sector of a free partition slot. Note that this information is already reported by the devices and it is used in the *space policy* dialog for sorting partitions.

### Getting the Resulting Config

The resulting "solved" config can be automatically generated from the probed and the staging devicegraphs. For example, let say we have resized *vda1*, deleted *vda2* and created a new partition. The resulting config would look like the following:

```json
{
  "drives": [
    {
      "search": "/dev/vda",
      "partitions": [
        {
          "search": "/dev/vda1",
          "size": "2 GiB"
        },
        {
          "search": "/dev/vda2",
          "delete": true
        }
        {  
          "sectorStart": 12442323342,
          "size": "10 GiB",
          "filesystem": {
            "type": "xfs",
            "path": "/home"
          }
        }
      ]
    }
  ]
}
```

**Comment by ancorgs** [comment]
We are currently already implementing a user interface based on the alternative presented here as "idea 2".
It's even documented (screenshots included) at https://agama-project.github.io/docs/user/interactive/storage#overall-description-of-the-storage-page

Additionally, we explored "idea 3" on this Hack Week project. https://hackweek.opensuse.org/24/projects/agama-expert-partitioner

 I'm closing the discussion for now.
